File: 42-trapping-rain-water/42-trapping-rain-water.scala
import scala.collection.mutable._

object Solution {
    def trap(height: Array[Int]): Int = {
        val stack: Stack[(Int, Int)] = Stack()
        var res = 0

        height.zipWithIndex.foreach { case (h, idx) =>
            while(stack.nonEmpty && stack.top._2 < h) {
                val (pidx, ph) = stack.pop
                val d = if (stack.isEmpty) 0 else (Math.min(stack.top._2, h) - ph)
                val w = if (stack.isEmpty) 0 else idx - stack.top._1 - 1
                res += d * w
            }

            if (stack.isEmpty || stack.top._2 >= h) stack.push((idx, h))
        }

        res
    }
}

File: 42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int res = 0;
        
        Deque<int[]> stack = new LinkedList<>();
        
        for(int i = 0; i < height.length; i++) {
            int h = height[i];
            
            while(!stack.isEmpty() && stack.peek()[0] < h) {
                int[] prev = stack.pop();
                if (stack.isEmpty()) continue;
                res += (Math.min(stack.peek()[0], h) - prev[0]) * (i - stack.peek()[1] - 1);
            }
            
            stack.push(new int[] {h, i});
        }
        
        return res;
    }
}